pkg/provider/senhasegura/provider.go

```diff
@@ -0,0 +1,99 @@
+/*
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package senhasegura
+
+import (
+	"context"
+	"fmt"
+
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	esv1beta1 "github.com/external-secrets/external-secrets/apis/externalsecrets/v1beta1"
+	senhaseguraAuth "github.com/external-secrets/external-secrets/pkg/provider/senhasegura/auth"
+	"github.com/external-secrets/external-secrets/pkg/provider/senhasegura/dsm"
+)
+
+// https://github.com/external-secrets/external-secrets/issues/644
+// https://github.com/external-secrets/external-secrets/pull/967
+var _ esv1beta1.Provider = &Provider{}
+
+// Provider struct that satisfier ESO interface.
+type Provider struct{}
+
+const (
+	errUnknownProviderService = "unknown senhasegura Provider Service: %s"
+	errNilStore               = "nil store found"
+	errMissingStoreSpec       = "store is missing spec"
+	errMissingProvider        = "storeSpec is missing provider"
+	errInvalidProvider        = "invalid provider spec. Missing senhasegura field in store %s"
+)
+
+/*
+	Construct a new secrets client based on provided store
+*/
+func (p *Provider) NewClient(ctx context.Context, store esv1beta1.GenericStore, kube client.Client, namespace string) (esv1beta1.SecretsClient, error) {
+	err := p.ValidateStore(store)
+	if err != nil {
+		return nil, err
+	}
+
+	spec := store.GetSpec()
+	provider := spec.Provider.Senhasegura
+
+	isoSession, err := senhaseguraAuth.Authenticate(ctx, store, provider, kube, namespace)
+	if err != nil {
+		return nil, err
+	}
+
+	if provider.Module == esv1beta1.SenhaseguraModuleDSM {
+		return dsm.New(isoSession)
+	}
+
+	return nil, fmt.Errorf(errUnknownProviderService, provider.Module)
+}
+
+// Validate store using Validating webhook during secret store creating
+// Checks here are usually the best experience for the user, as the SecretStore will not be created until it is a 'valid' one.
+// https://github.com/external-secrets/external-secrets/pull/830#discussion_r833278518
+func (p *Provider) ValidateStore(store esv1beta1.GenericStore) error {
```

Comment:
You can base this method on the implementation for other providers. You can look at AWS SM or the 1Pass PR that is open.

Reply:
After removing ClientID and URL from the secret on the base types, we can `url.Parse` the URL to check if it is a valid URL containing `http://` or `https://`.

Reply:
Done!

Note: senhasegura only accepts requests using HTTPS for security reasons, i enforced in checks that provider URL uses https scheme, no other will be accepted in checks